<div style="background-color: lightblue;border: 2px solid gray;">
    <h2 style ="text-align:center; padding-top:5px;"> CS 101 - Foundation of Data Science and Engineering  </h2><br>
    <p style="text-align:center;padding:5px; fontt-size:14px"><b> PSET-5 - Exploratory Data Analysis (100 pts)<b></p> <br>
</div>

# This is an individual assignment. No collaboration is allowed!

### Note: 
You are allowed to use all Python built-in functions and other Import features covered in class. Ensure your code is organized, well-commented, and follows the best practices we discussed. Remember, the key is not just to write a working program but to produce a solution that follows SPECS, is easy to debug, and is easy to maintain. 

### Assignment Question List:
 
**Question 1 : Data Extraction and cleaning (1 pts)**

**Question 2 : Column Cleaning Function (15 pts)**

**Question 3 : Data Cleaning (9 pts)**

**Question 4 : Data Aggregation (30 pts)**

**Question 5 : Merging Data (10 pts)**

**Question 6 : Filtering Data (30 pts)**

**Coding Style : (5 pts)**

## Objective:
**You are tasked with performing detailed exploratory data analysis on various system datasets: CPU, Disk, and Memory. Utilize Python functions to streamline data extraction, cleaning, and aggregation.**



### Datasets:

1. **CPU Dataset: `cpu.csv`**
   - **Columns:** 'Image', 'PID', 'Description', 'Status', 'Threads', 'CPU', 'Average CPU'
   <br><br>
2. **Disk Dataset: `disk.csv`**
   - **Columns:** 'Image', 'PID', 'Model', 'Read Byte Per Second', 'Write Byte Per Second', 'Delay', 'Total Byte Per Second'
    <br><br>
3. **Memory Dataset: `memory.csv`**
   - **Columns:** 'Image', 'PID', 'Hard Faults/sec', 'Commit KB', 'Working Set KB', 'Shareable KB', 'Private KB'
    <br><br>

# Your Tasks

### 1. Data Extraction (1 pts):

a. **Load `cpu.csv` into a Pandas DataFrame named `df_cpu`. Extract only these columns:** 'Image', 'PID', 'Description', 'Status', 'Threads', 'Average CPU'

b. **Load `disk.csv` into a DataFrame named `df_disk`. Extract only these columns:** 'Image', 'PID', 'Total Byte Per Second'

c. **Load `memory.csv` into a DataFrame named `df_memory`. Extract only these columns:** 'Image', 'PID', 'Working Set KB'

d. **For each of the above data frame show the shape, and info(example : df_cpu.info())**

In [108]:
import pandas as pd

#load datasets
df_cpu = pd.read_csv("cpu.csv")
df_disk = pd.read_csv("disk.csv")
df_memory = pd.read_csv("memory.csv")



In [109]:
df_cpu.shape, df_cpu.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306 entries, 0 to 305
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Image        306 non-null    object 
 1   PID          306 non-null    object 
 2   Description  290 non-null    object 
 3   Status       306 non-null    object 
 4   Threads      306 non-null    object 
 5   CPU          306 non-null    int64  
 6   Average CPU  306 non-null    float64
dtypes: float64(1), int64(1), object(5)
memory usage: 16.9+ KB


((306, 7), None)

In [110]:
df_disk.shape, df_disk.info(), 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 7 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   Image                  14 non-null     object
 1   PID                    14 non-null     int64 
 2   Model                  14 non-null     object
 3   Read Byte Per Second   14 non-null     object
 4   Write Byte Per Second  14 non-null     object
 5   Delay                  14 non-null     int64 
 6   Total Byte Per Second  14 non-null     object
dtypes: int64(2), object(5)
memory usage: 916.0+ bytes


((14, 7), None)

In [111]:

df_memory.shape, df_memory.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306 entries, 0 to 305
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Image            306 non-null    object
 1   PID              306 non-null    int64 
 2   Hard Faults/sec  306 non-null    int64 
 3   Commit KB        306 non-null    object
 4   Working Set KB   306 non-null    object
 5   Shareable KB     306 non-null    object
 6   Private KB       306 non-null    object
dtypes: int64(2), object(5)
memory usage: 16.9+ KB


((306, 7), None)

### 2. Column Cleaning Function (15 pts):

a. **Write a Python function, `CleanColumnHeading(dfx)`, to clean the column headers of any Pandas dataframe. The function should be dynamic enough to be able to process any datasets. The function should:**

   i. Convert all column names to lowercase.
   
   ii. Replace spaces in column names with underscores `_`.

   iii. Apply the `CleanColumnHeading` function to `df_cpu`, `df_disk`, and `df_memory`.

   iv. Show examples of the changes.  

**Example:**

Function header:
```python
def CleanColumnHeading(dfx):
    # Your code to convert all column names to lowercase
    # Your code to change all spaces in column names to underscores _
    return dfx


In [112]:
def CleanColumnHeading(dfx):
    dfx.columns = dfx.columns.str.strip().str.lower().str.replace(' ', '_')
    return dfx


In [113]:
df_cpu = CleanColumnHeading(df_cpu)
df_disk = CleanColumnHeading(df_disk)
df_memory = CleanColumnHeading(df_memory)

In [114]:
df_cpu.info()
print(df_cpu.head(3))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306 entries, 0 to 305
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   image        306 non-null    object 
 1   pid          306 non-null    object 
 2   description  290 non-null    object 
 3   status       306 non-null    object 
 4   threads      306 non-null    object 
 5   cpu          306 non-null    int64  
 6   average_cpu  306 non-null    float64
dtypes: float64(1), int64(1), object(5)
memory usage: 16.9+ KB
            image    pid  description     status threads  cpu  average_cpu
0   Secure System    140          NaN  Suspended       -    0          0.0
1  SearchHost.exe   9272   SearchHost  Suspended      68    0          0.0
2     LockApp.exe  12784  LockApp.exe  Suspended      16    0          0.0


In [115]:
df_disk.info()
print(df_disk.head(3))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 7 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   image                  14 non-null     object
 1   pid                    14 non-null     int64 
 2   model                  14 non-null     object
 3   read_byte_per_second   14 non-null     object
 4   write_byte_per_second  14 non-null     object
 5   delay                  14 non-null     int64 
 6   total_byte_per_second  14 non-null     object
dtypes: int64(2), object(5)
memory usage: 916.0+ bytes
        image    pid model read_byte_per_second write_byte_per_second  delay  \
0  chrome.exe   5448     x                    0                 6,278      0   
1  chrome.exe  13324    x3                    0                 4,940      0   
2   EXCEL.EXE  12132    x1               71,807                     0      0   

  total_byte_per_second  
0                 6,278  
1          

In [116]:
df_memory.info()
print(df_memory.head(3))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306 entries, 0 to 305
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   image            306 non-null    object
 1   pid              306 non-null    int64 
 2   hard_faults/sec  306 non-null    int64 
 3   commit_kb        306 non-null    object
 4   working_set_kb   306 non-null    object
 5   shareable_kb     306 non-null    object
 6   private_kb       306 non-null    object
dtypes: int64(2), object(5)
memory usage: 16.9+ KB
                           image    pid  hard_faults/sec commit_kb  \
0  AcrobatNotificationClient.exe  18096                0    40,272   
1      Adobe Crash Processor.exe   7856                0     4,768   
2      Adobe Desktop Service.exe   6640                0    85,284   

  working_set_kb shareable_kb private_kb  
0          3,632        3,392        240  
1         15,252       13,704      1,548  
2         85,504       75,696 

### 3. Data cleaning (9 pts)
**Examine the columns 'threads', 'total_byte_per_second', 'working_set_kb' from the dataframes df_cpu, df_disk, and df_memory. We are going to work with these columns Question 4-6. Ensure that they have the correct data type, fix the values if required so. If there are invalid values drop them. At the end of this step you should not have any invalid values, and the correct data type set for the 3 columns. You are not required to do data cleaning on other columns. If you choose to do so, please ensure that no rows are dropped while cleaning these columns** 

**Hint: values such as 71,807 are not invalid. They are simply string representation of the number 71807. Fix it so they are stored as 71807, and the column datatype should either be a float or int.** 


**For each of the columns modified , show example rows of what was modified. Also call the info() method on each of the dataframes.**


In [117]:
def clean_numeric_column(dfx, column_name):
    dfx[column_name] = pd.to_numeric(dfx[column_name].astype(str).str.replace(',', ''), errors='coerce')

    #drop rows with NaN
    dfx.dropna(subset=[column_name], inplace=True)
    #convert to integer type if possible
    if dfx[column_name].dtype == 'float64' and (dfx[column_name] % 1 == 0).all():
        dfx[column_name] = dfx[column_name].astype(int)

    return dfx

In [118]:
df_cpu_beforecleaned = df_cpu.copy()
df_disk_beforecleaned = df_disk.copy()
df_memory_beforecleaned = df_memory.copy()
df_cpu = clean_numeric_column(df_cpu, 'threads')
df_disk = clean_numeric_column(df_disk, 'total_byte_per_second')
df_memory = clean_numeric_column(df_memory, 'working_set_kb')


In [119]:
#identify modified
df_cpu_modified = df_cpu_beforecleaned[
    df_cpu_beforecleaned['threads'].astype(str).str.contains(',', na=False)]
df_disk_modified = df_disk_beforecleaned[
    df_disk_beforecleaned['total_byte_per_second'].astype(str).str.contains(',', na=False)]
df_memory_modified = df_memory_beforecleaned[
    df_memory_beforecleaned['working_set_kb'].astype(str).str.contains(',', na=False)]

In [120]:
df_cpu.info()

print("Examples of modified values in df_cpu in 'threads' column:")
if df_cpu_modified.empty:
    print("No modifications were needed in 'threads' column.")
else:
    print(df_cpu_modified.head())



<class 'pandas.core.frame.DataFrame'>
Index: 304 entries, 1 to 305
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   image        304 non-null    object 
 1   pid          304 non-null    object 
 2   description  289 non-null    object 
 3   status       304 non-null    object 
 4   threads      304 non-null    int64  
 5   cpu          304 non-null    int64  
 6   average_cpu  304 non-null    float64
dtypes: float64(1), int64(2), object(4)
memory usage: 19.0+ KB
Examples of modified values in df_cpu in 'threads' column:
No modifications were needed in 'threads' column.


In [121]:
df_disk.info()

print("Examples of modified values in df_disk in 'total_byte_per_second' column:")
print(df_disk_modified.head())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 7 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   image                  14 non-null     object
 1   pid                    14 non-null     int64 
 2   model                  14 non-null     object
 3   read_byte_per_second   14 non-null     object
 4   write_byte_per_second  14 non-null     object
 5   delay                  14 non-null     int64 
 6   total_byte_per_second  14 non-null     int64 
dtypes: int64(3), object(4)
memory usage: 916.0+ bytes
Examples of modified values in df_disk in 'total_byte_per_second' column:
                   image    pid model read_byte_per_second  \
0             chrome.exe   5448     x                    0   
1             chrome.exe  13324    x3                    0   
2              EXCEL.EXE  12132    x1               71,807   
3  Grammarly.Desktop.exe  19180    h1                6,729   


In [122]:
df_memory.info()
print("Examples of modified values in df_memory in 'working_set_kb' column:")
print(df_memory_modified.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306 entries, 0 to 305
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   image            306 non-null    object
 1   pid              306 non-null    int64 
 2   hard_faults/sec  306 non-null    int64 
 3   commit_kb        306 non-null    object
 4   working_set_kb   306 non-null    int64 
 5   shareable_kb     306 non-null    object
 6   private_kb       306 non-null    object
dtypes: int64(3), object(4)
memory usage: 16.9+ KB
Examples of modified values in df_memory in 'working_set_kb' column:
                           image    pid  hard_faults/sec commit_kb  \
0  AcrobatNotificationClient.exe  18096                0    40,272   
1      Adobe Crash Processor.exe   7856                0     4,768   
2      Adobe Desktop Service.exe   6640                0    85,284   
3            AdobeCollabSync.exe  17928                0     7,016   
4            AdobeCol

### 4. Data Aggregation (30 pts):

a. **Develop an `aggregate_data(dfx)` function that:**

   - Accepts one of the dataframes (`df_cpu`, `df_disk`, or `df_memory`).
   - Dynamically identifies the dataset based on its columns.
   - Aggregates the data based on the specific dataset requirements provided below and returns an aggregated dataframe. You can name this new aggregated dataframe whatever name works (examples given).

b. **Aggregation Requirements:**

   i. For `df_cpu`:
   
          1. Group by 'image' and aggregate the 'threads' column. Use the sum() function for aggregation.
          2. The resultant dataframe should contain: 'image_name', 'process_qty', 'threads_sum'.
          3. Reset the index after aggregation.
          4. Assign the new aggregated dataframe values to df_cpu_sum.
   
   ii. For `df_disk`:
   
          1. Group by 'image' and aggregate the 'total_byte_per_second' column.Use the sum() function for aggregation.
          2. The resultant dataframe should contain: 'image_name', 'process_qty', 'total_byte_sum'.
          3. Reset the index after aggregation.
          4. Assign the new aggregated dataframe values to df_disk_sum.
   
   iii. For `df_memory`:
   
          1. Group by 'image' and aggregate the 'working_set_kb' column. Use the sum() function for aggregation.
          2. The resultant dataframe should contain: 'image_name', 'process_qty', 'working_set_sum'.
          3. Reset the index after aggregation.
          4. Assign the new aggregated dataframe values to df_memory_sum.
   
   **Note: 'process_qty' is defined by grouping by 'image' and determining the number of times a particular image name occurs. Use the size() function for aggregation**
          
c. **Run `aggregate_data()` for each of your dataframes and save them into new dataframes. Print the head(10) and tail(10) for each of the new dataframes.**

**Example:**

Function header:
```python
def aggregate_data(dfx):
    # Your code aggregates the data based on the requirements provided.
    return df_combo  # Name this resultant dataframe whatever name that works
```
Calling function and assigning the resultant dataframe to `df_cpu_sum`, `df_disk_sum`, `df_memory_sum`:
```python
df_cpu_sum = aggregate_data(df_cpu)
df_disk_sum = aggregate_data(df_disk)
df_memory_sum = aggregate_data(df_memory)
```


In [123]:
def aggregate_data(dfx):
    if 'threads' in dfx.columns:
        # Aggregating df_cpu
        df_combo = dfx.groupby('image', as_index=False).agg(
            process_qty=('image', 'size'),
            threads_sum=('threads', 'sum')
        )
    elif 'total_byte_per_second' in dfx.columns:
        # Aggregating df_disk
        df_combo = dfx.groupby('image', as_index=False).agg(
            process_qty=('image', 'size'),
            total_byte_sum=('total_byte_per_second', 'sum')
        )
    elif 'working_set_kb' in dfx.columns:
        # Aggregating df_memory
        df_combo = dfx.groupby('image', as_index=False).agg(
            process_qty=('image', 'size'),
            working_set_sum=('working_set_kb', 'sum')
        )
    return df_combo

In [124]:
df_cpu_sum = aggregate_data(df_cpu)

print("Aggregated df_cpu_sum:")
print(df_cpu_sum.head(10))
print(df_cpu_sum.tail(10))


Aggregated df_cpu_sum:
                           image  process_qty  threads_sum
0  AcrobatNotificationClient.exe            1           13
1      Adobe Crash Processor.exe            1            4
2      Adobe Desktop Service.exe            1           47
3            AdobeCollabSync.exe            2           24
4             AdobeIPCBroker.exe            1           26
5    AdobeNotificationClient.exe            1           12
6         AdobeUpdateService.exe            1            5
7             AggregatorHost.exe            1            1
8       ApplicationFrameHost.exe            1            6
9                  CCLibrary.exe            1            1
                          image  process_qty  threads_sum
140  svchost.exe (osprivacy -p)            1            4
141       svchost.exe (smphost)            1            6
142     svchost.exe (utcsvc -p)            1           12
143               taskhostw.exe            1            8
144                  uihost.exe       

In [125]:
df_disk_sum = aggregate_data(df_disk)

print("Aggregated df_disk_sum:")
print(df_disk_sum.head(10))
print(df_disk_sum.tail(10))

Aggregated df_disk_sum:
                   image  process_qty  total_byte_sum
0              EXCEL.EXE            1           71807
1  Grammarly.Desktop.exe            1            6729
2            MsMpEng.exe            1            1457
3            OUTLOOK.EXE            1            1471
4               Registry            1            5523
5                 System            1          232627
6              Teams.exe            1             547
7           WUDFHost.exe            1           24587
8             chrome.exe            2           11218
9     msedgewebview2.exe            4           20060
                   image  process_qty  total_byte_sum
0              EXCEL.EXE            1           71807
1  Grammarly.Desktop.exe            1            6729
2            MsMpEng.exe            1            1457
3            OUTLOOK.EXE            1            1471
4               Registry            1            5523
5                 System            1          232627
6   

In [126]:
df_memory_sum = aggregate_data(df_memory)

print("Aggregated df_memory_sum:")
print(df_memory_sum.head(10))
print(df_memory_sum.tail(10))

Aggregated df_memory_sum:
                           image  process_qty  working_set_sum
0  AcrobatNotificationClient.exe            1             3632
1      Adobe Crash Processor.exe            1            15252
2      Adobe Desktop Service.exe            1            85504
3            AdobeCollabSync.exe            2            37012
4             AdobeIPCBroker.exe            1            10568
5    AdobeNotificationClient.exe            1             2836
6         AdobeUpdateService.exe            1             9580
7             AggregatorHost.exe            1             7420
8       ApplicationFrameHost.exe            1            35988
9                  CCLibrary.exe            1             2948
                          image  process_qty  working_set_sum
141  svchost.exe (osprivacy -p)            1            13812
142       svchost.exe (smphost)            1            16624
143     svchost.exe (utcsvc -p)            1            30104
144               taskhostw.exe  

### 5. Merging Data (10 pts):

a. **Use the Pandas merge function to do an inner join on 'image_name' for dataframes `df_cpu_sum` and `df_disk_sum` created in the previous step. Store the result in a dataframe named `df_new`**.

b. **Further merge `df_new` with `df_memory_sum` using an inner join on 'image_name'**.

c. **Show the resulting dataframe in each of the steps above**

Please refer to the pandas merge method and its parameters here : https://pandas.pydata.org/docs/reference/api/pandas.merge.html


In [127]:
df_new = pd.merge(df_cpu_sum, df_disk_sum, on='image', how='inner')

print("Merge of df_cpu_sum and df_disk_sum:")
print(df_new.head(10))

Merge of df_cpu_sum and df_disk_sum:
                   image  process_qty_x  threads_sum  process_qty_y  \
0              EXCEL.EXE              1           64              1   
1  Grammarly.Desktop.exe              1           48              1   
2            MsMpEng.exe              1           66              1   
3            OUTLOOK.EXE              1           89              1   
4               Registry              1            4              1   
5                 System              1          342              1   
6              Teams.exe              9          213              1   
7           WUDFHost.exe              6           66              1   
8             chrome.exe             12          236              2   
9     msedgewebview2.exe             32          620              4   

   total_byte_sum  
0           71807  
1            6729  
2            1457  
3            1471  
4            5523  
5          232627  
6             547  
7           24587  
8

In [128]:
df_new = pd.merge(df_new, df_memory_sum, on='image', how='inner')

print("Merge with df_memory_sum:")
print(df_new.head(10))

Merge with df_memory_sum:
                   image  process_qty_x  threads_sum  process_qty_y  \
0              EXCEL.EXE              1           64              1   
1  Grammarly.Desktop.exe              1           48              1   
2            MsMpEng.exe              1           66              1   
3            OUTLOOK.EXE              1           89              1   
4               Registry              1            4              1   
5                 System              1          342              1   
6              Teams.exe              9          213              1   
7           WUDFHost.exe              6           66              1   
8             chrome.exe             12          236              2   
9     msedgewebview2.exe             32          620              4   

   total_byte_sum  process_qty  working_set_sum  
0           71807            1           298380  
1            6729            1           273436  
2            1457            1           2

### 6. Filtering Data (30 pts):

a. **Filter `df_new` to obtain:**

   i. `image_name` that had 'working_set_sum' greater than 200,000. Store the result in a variable named 'high_memory_image'.

   ii. `image_name` that had 'thread_sum' greater than 200. Store the result in a variable named 'high_thread_image'.

   iii. `image_name` that had 'working_set_sum' greater than 200,000, 'thread_sum' less than 50, and 'total_byte_sum' less than 7,000. Store the result in a variable named 'hi_mem_low_thread_low_io'.
   
   iv. Show each of the filtered dataframe in separate cells. Feel free to add new cells to this notebook.  
   


In [129]:
#filter images where 'working_set_sum' is greater than 200,000
high_memory_image = df_new[df_new['working_set_sum'] > 200000]

print("high memory usage (working_set_sum > 200,000):")
print(high_memory_image.head(10))

high memory usage (working_set_sum > 200,000):
                   image  process_qty_x  threads_sum  process_qty_y  \
0              EXCEL.EXE              1           64              1   
1  Grammarly.Desktop.exe              1           48              1   
2            MsMpEng.exe              1           66              1   
3            OUTLOOK.EXE              1           89              1   
6              Teams.exe              9          213              1   
8             chrome.exe             12          236              2   
9     msedgewebview2.exe             32          620              4   

   total_byte_sum  process_qty  working_set_sum  
0           71807            1           298380  
1            6729            1           273436  
2            1457            1           210204  
3            1471            1           447040  
6             547           10          1117452  
8           11218           12          1076912  
9           20060           32    

In [130]:
#filter images where 'threads_sum' is greater than 200
high_thread_image = df_new[df_new['threads_sum'] > 200]

print("high thread count (threads_sum > 200):")
print(high_thread_image.head(10))

high thread count (threads_sum > 200):
                image  process_qty_x  threads_sum  process_qty_y  \
5              System              1          342              1   
6           Teams.exe              9          213              1   
8          chrome.exe             12          236              2   
9  msedgewebview2.exe             32          620              4   

   total_byte_sum  process_qty  working_set_sum  
5          232627            1             3140  
6             547           10          1117452  
8           11218           12          1076912  
9           20060           32          1002516  


In [131]:
hi_mem_low_thread_low_io = df_new[
    (df_new['working_set_sum'] > 200000) & 
    (df_new['threads_sum'] < 50) & 
    (df_new['total_byte_sum'] < 7000)
]

print("high memory, low thread, and low I/O:")
print(hi_mem_low_thread_low_io.head(10))

high memory, low thread, and low I/O:
                   image  process_qty_x  threads_sum  process_qty_y  \
1  Grammarly.Desktop.exe              1           48              1   

   total_byte_sum  process_qty  working_set_sum  
1            6729            1           273436  


### Coding Style (5 pts) 
Although we do not enforce a coding style such as PEP 8 (https://peps.python.org/pep-0008/) , please ensure that you have comments for each of the functions defined. Your code is readable, and includes only the code that is required by the assignment. Please remove any commented code, and experimental code that you may have tried. For each of the questions be sure to show some example rows of the dataframe that was modified or created.  


## Submission on Gradescope

**Gradescope canvas left menu -> Gradescop -> PSET 5: Exploratory Data Analysis**

**Submission :**
Submit the jupyter notebook, and a pdf version of this notebook.

To create a pdf of this notebook :  In your browser open print, and save as pdf. Name the pdf LastNameFirstName_pset5.pdf
example: DoeJohn_pset5.pdf

Name this jupyter notebook with the same format LastNameFirstName_pset5.ipynb

Make sure that your notebook has been run before creating pdf. Any outputs from running the code needs to be clearly visible. We need both .ipynb, and pdf of this notebook to assign you grades. 

Drop all the files in gradescope under PSET 5: Exploratory Data Analysis.
